src/main/java/graphql/introspection/IntrospectionUtils.java

```diff
@@ -0,0 +1,32 @@
+package graphql.introspection;
+
+import graphql.PublicApi;
+
+@PublicApi
+public class IntrospectionUtils {
+    public static final String INTROSPECTION_SCHEMA_STRING = "__schema";
+
+    /**
+     * Tests whether the provided {@code query} represents a GraphQL introspection query or not.  Whether or not
+     * the query is the introspection query is determined by the following in the order given:
+     *
+     * 1.  If the operation name is equal {@link IntrospectionQuery#INTROSPECTION_QUERY_NAME} or
+     * 2.  If query contains operation name {@link IntrospectionQuery#INTROSPECTION_QUERY_NAME} or
+     * 3.  If query contains {@link IntrospectionUtils#INTROSPECTION_SCHEMA_STRING}
+     *
+     * @param operationName The operation name of the query.
+     * @param query         Optional query to test.  If provided, and the operationName is not
+     *                      IntrospectionQuery.INTROSPECTION_QUERY_NAME, then query is scanned to check for the
+     *                      name or the existence of __schema
+     *
+     * @return {@code true} if the operation name or query represent the introspection query, false otherwise.
+     */
+    public static boolean isIntrospectionQuery(final String operationName, final String query) {
+        boolean isIntrospectionQueryName = IntrospectionQuery.INTROSPECTION_QUERY_NAME.equals(operationName);
+
+        return isIntrospectionQueryName ||
+            ((null != query) &&
+             !query.isEmpty() &&
+             (query.contains(IntrospectionQuery.INTROSPECTION_QUERY_NAME) || query.contains(IntrospectionUtils.INTROSPECTION_SCHEMA_STRING)));
+    }
+}
```

Comment:
I dont think the first two really help.  For example

````
   query Introspection {
        thisIsNot {
          actuallyIntrospection
       }
   }
```

Why not just do `query.contains(IntrospectionUtils.INTROSPECTION_SCHEMA_STRING)`

Reply:
Yes, the other options are not fool proof.  I had some concerns about possibly running into performance issues with the case that scans for the IntrospectionUtils.INTROSPECTION_SCHEMA_STRING in scenarios where you have a long query and it's not the introspection query.  So, I was trying to use the name as a means of giving developers a quick way of checking at the risk of having less accuracy.

Our use case for using this utility is to avoid running some metrics instrumentation code when the query is the introspection query.  I'm going to run some tests to check the performance impact for different scenarios.  Hopefully, they show minimal differences between the different scenarios and I can just change the code to what you suggest.

Reply:
I ran some performance numbers for this utility method, see test details below.  I don't think my above utility method will be very performant since I expect that most operations run will not be the introspection query and if I just use query.contains(IntrospectionUtils.INTROSPECTION_SCHEMA_STRING) as the check, then the longer the query is, the less performant the method will be.  This was the reasoning behind passing the in the operation name as a quick way to short circuit the check; although it has the potential for false-positives if misused or operation names are misused.

What I really want to do though is not execute certain instrumentation code when something is the introspection query but I want to be able to make this check in a performant way which I don't think this method will actually do.

Does graphql-java have a way to know if a particular field is an introspection field?  Our instrumentation logs metrics and we do not want to log metrics for fields tied to trivial data fetchers or for instrumentation-based fields.  However, instrumentation-based fields are not currently considered to have trivial data fetchers.  The other thing is that not all instrumentation beginField methods give you access to isTrivialDataFetcher check that the beginFieldFetch instrumentation parameters provides.  We've found a workaround for the trivial data fetcher check but not for the instrumentation check.  So, if you have any other suggestions on how this can be accomplished, I'm open to them.  Maybe a different PR change to help in this would be required?

In my tests, I used the following scenario:
Loops: 1,000,000

Case 1:  Just an empty loop for a baseline

Case 2:
* Operation Name: IntrospectionQuery
* Query: IntrospectionQuery.INTROSPECTION_QUERY (the actual introspection query as defined by graphql-java

Case 3:
* Operation Name: IntrospectionQuery
* Query: A variation of IntrospectionQuery.INTROSPECTION_QUERY which changes the operation name with  "NonIntrospectionQueryName" and __schema with just "schema" so the query is as long as the real introspection query but will return false for the query checks.  I'll refer to this as the NonIntrospectionQuery from now on.

Case 4:
* Operation Name: NonIntrospectionQueryName
* Query: IntrospectionQuery.INTROSPECTION_QUERY

Case 5:
* Operation Name: NonIntrospectionQueryName
* Query: NonIntrospectionQuery

IntrospectionQuery.INTROSPECTION_QUERY

Results:
```
Case 1:  1000000 Loops Empty Loop: time: 25.537075 ms, avg: 0.000025537075 ms
Case 2:  1000000 Loops IntrospectionQuery By Name (name: true, query: true): time: 118.105139 ms, avg: 0.000118105139 ms
Case 3:  1000000 Loops IntrospectionQuery By Name (name: true, query: false): time: 115.691338 ms, avg: 0.000115691338 ms
Case 4:  1000000 Loops IntrospectionQuery By Query (name: false, query: true): time: 113.963149 ms, avg: 0.000113963149 ms
Case 5:  1000000 Loops Non-IntrospectionQuery By Query (name: false, query: false): time: 1099.623745 ms, avg: 0.001099623745 ms
```

Reply:
Would it help if we added another utility that interacted with the parsed `Document`? Maybe in:
```
InstrumentationContext<Document> beginParse(InstrumentationExecutionParameters parameters);
```
Maybe there is a way to get the root `Document` node and determine if it's `__schema`? Could also keep the `String` one above with the performance caveat noted?

Reply:
I'll have a look at that.  I was hoping that graphql-java could somehow have some of that information already in the document so that we can avoid walking the document in instrumentation calls, but let me check if that is possible.  Like you state @brockgibson, this utility might still be useful in some cases as long as we note it's performance impact.

Reply:
It seems that in instrumentation, in some methods I have access to the GraphQL Document.  I can use that to check for an __schema node.  However, since fragments can be used, I believe I must convert some form of the document's selection set into a FieldSubSelection which would already have the fragments applied.  Looking at some of the code in graphql-java, I'm not clear if this process actually performs any data resolution of the fields because I don't actually want that to happen.  Most of the code doing this translation also takes in a GraphQLSchema, an GraphQLObjectType, the fragments, variables, etc.  I'm wondering, how many of those are actually necessary/required if all I want is to produce a FieldSubSelection with the fragments taken into account so I have the actual field names?  Hopefully, this will just work by just needing the fragments and the OperationDefinition node found in the Document